### PR TITLE
GG-37466 Fix handling lost partitions after cluster restart

### PR DESCRIPTION
**Pull request by sk0x50**
https://ggsystems.atlassian.net/browse/GG-37466

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 1]
```diff
@@ -434,6 +434,9 @@ private void saveStoreMetadata(
 
                         if (part != null && part.state() != GridDhtPartitionState.EVICTED)
                             state = part.state();
```
Comment (vldpyatkov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:436]: Why do you not write right here?
`state = part.state() == LOST ? OWNING : part.state()`

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java [hunk 2]
```diff
@@ -162,6 +175,142 @@ public void testConsistencyAfterResettingLostPartitions_3() throws Exception {
...
+            cache.put(k, k);
+
+        // Stop the second node and update data in the cache to trigger rebalancing when the node will be restarted.
+        stopGrid(1);
+
+        int owningPartId = 12;
+
+        // Rewrite data to trigger further rebalance.
+        for (int k = 2500; k < 5000; k++) {
+            if (!stopInTheMiddleOfCheckpoint) {
```
Comment (vldpyatkov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java:223]: I would not use a negative over a whole condition in a clause with else.

**Review by vldpyatkov — APPROVED** [review]

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java [hunk 3]
```diff
@@ -162,6 +175,142 @@ public void testConsistencyAfterResettingLostPartitions_3() throws Exception {
...
+
+            checkLostPartitionAcrossCluster(DEFAULT_CACHE_NAME, partId);
+        }
+
+        assertEquals(
+            "WAL should be " + (stopInTheMiddleOfCheckpoint ? "enabled" : "disabled") + " for the cache + " + DEFAULT_CACHE_NAME,
+            stopInTheMiddleOfCheckpoint,
+            ig1.context().cache().cacheGroup(CU.cacheId(DEFAULT_CACHE_NAME)).localWalEnabled());
+
+        if (!stopInTheMiddleOfCheckpoint)
```
Comment (vldpyatkov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java:261]: I would not use a negative over a whole condition in a clause with else.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java (hunk 1 above)
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java]: Done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java (hunk 2 above)
Comment (sk0x50) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java]: Ok. Fixed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java (hunk 3 above)
Comment (sk0x50) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/CachePartitionLossWithPersistenceTest.java]: Fixed

**Pull request merged**